MAINT: Remove Python <3.8 support from C

seberg opened the pull request:
Remove version switches for support of Python versions before 3.8.
The biggest deletion is that we can use Pythons `PyVectorcall_Call`
to support the old `*args, **kwargs` style call convention in ufuncs.

seberg committed:
MAINT: Remove Python <3.8 support from C

Remove version switches for support of Python versions before 3.8.
The biggest deletion is that we can use Pythons `PyVectorcall_Call`
to support the old `*args, **kwargs` style call convention in ufuncs.

WarrenWeckesser commented:
Apparently LGTM is using Python 3.7.

charris commented:
> Apparently LGTM is using Python 3.7.

They were supposed to fix that last quarter, we will have to disable LGTM if they don't. Supposedly there is a more current replacement (code scanning alerts) in the security tab we can use as a replacement.

WarrenWeckesser commented:
For reference, here's the mailing list thread from August: [Drop LGTM testing.](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/IXZIQHHX2HJ235NTPWRBOH2OE3YPBIHU/)

WarrenWeckesser commented:
Let's just drop LGTM.  We can restore it later if LGTM updates their python.

Does it make sense to do that here, or would it be cleaner to do it in a separate PR?

charris commented:
> Does it make sense to do that here,

We can disable the access or remove it in `settings -- installed github apps`. Let's turn off access and try the github code scanner that is based on it. Github acquired Semle a while ago.

charris commented:
I suspended the app. Let's restart and see what happens.

charris commented:
I've added the github code scanner, it should be using Python 3.8 (Ubuntu 20-04). If not we can remove it again :)

WarrenWeckesser commented:
Looks like CodeQL does a shallow checkout of the numpy git repository.  That won't work, because versioneer needs the latest git tag to create the numpy version.

charris commented:
> Looks like CodeQL does a shallow checkout

That's not the only problem. I let it make the automatic commit, which was a mistake. I'm going to remove it and leave it for another day as it is going to take some effort to get it working and that is better done as a PR.

charris commented:
@seberg Could you also update the Python version check in setup.py?

seberg committed:
MAINT: Fix python version check in setup.py to require Python 3.8

Previously, this was not strictly enforced for the only purpose of
running LGTM (hoping it would be fixed before the 1.22 release).

charris commented:
Thanks Sebastian.  We could also update the default pickle protocol to 4, the default in Python 3.8.  That would allow large pickles by default. AFAICT, the current default is 2, probably for Python 2.7 compatibility. I'll make the proposition on the mailing list.

eric-wieser commented:
Is this a good idea? Doesn't this mean pip is unable to read the `python_requires` field in `setup.py`?

seberg commented:
@eric-wieser what you mean that it would break the PyPI listing?  https://pypi.org/project/numpy/1.22.0rc1/ does correctly show `Requires: Python >=3.8` in the side bar.
